src/modules/job-manager/event.c

```diff
@@ -596,12 +715,202 @@ int event_job_post_pack (struct event *event,
+        goto error;
+    }
+    zlist_freefn (event->listeners, el, events_listener_destroy, false);
+
+    return;
+
+error:
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
+    events_listener_destroy (el);
+}
+
+bool check_cancel (struct events_listener *el,
+                   uint32_t matchtag,
+                   const char *sender)
+{
+    uint32_t t;
+    char *s = NULL;
+    bool found = false;
+
+    if (!flux_msg_get_matchtag (el->request, &t)
+        && matchtag == t
+        && !flux_msg_get_route_first (el->request, &s)
+        && !strcmp (sender, s))
+        found = true;
+    free (s);
+    return found;
+}
+
+void events_cancel_request (flux_t *h, flux_msg_handler_t *mh,
+                            const flux_msg_t *msg, void *arg)
+{
+    struct job_manager *ctx = arg;
+    struct event *event = ctx->event;
+    struct events_listener *el;
+    uint32_t matchtag;
+    char *sender;
+
+    if (flux_request_unpack (msg, NULL, "{s:i}", "matchtag", &matchtag) < 0) {
+        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
+        return;
+    }
+    if (flux_msg_get_route_first (msg, &sender) < 0) {
+        flux_log_error (h, "%s: flux_msg_get_route_first", __FUNCTION__);
+        return;
+    }
+    el = zlist_first (event->listeners);
+    while (el) {
+        if (check_cancel (el, matchtag, sender))
+            break;
+        el = zlist_next (event->listeners);
+    }
+    if (el) {
+        if (flux_respond_error (h, el->request, ENODATA, NULL) < 0)
+            flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
+        zlist_remove (event->listeners, el);
+    }
+    free (sender);
+}
+
+void event_listeners_disconnect_rpc (flux_t *h,
+                                     flux_msg_handler_t *mh,
+                                     const flux_msg_t *msg,
+                                     void *arg)
+{
+    struct job_manager *ctx = arg;
+    struct event *event = ctx->event;
+    struct events_listener *el;
+    char *sender;
+    zlist_t *tmplist = NULL;
+
+    if (flux_msg_get_route_first (msg, &sender) < 0)
+        return;
+    el = zlist_first (event->listeners);
+    while (el) {
+        char *tmpsender;
+        if (flux_msg_get_route_first (el->request, &tmpsender) == 0) {
+            if (!strcmp (sender, tmpsender)) {
+                /* cannot remove from zlist while iterating, so we
+                 * store off entries to remove on another list */
+                if (!tmplist && !(tmplist = zlist_new ())) {
+                    errno = ENOMEM;
+                    flux_log_error (h, "%s: zlist_new", __FUNCTION__);
+                    goto error;
+                }
+                if (zlist_append (tmplist, el) < 0) {
+                    errno = ENOMEM;
+                    flux_log_error (h, "%s: zlist_append", __FUNCTION__);
+                    goto error;
+                }
```

Comment:
Distinguishing zlist_append failure from the zlist_new failure in logs is not helpful, and `tmpsender` is leaked if either fails.  Maybe a helper function?
```c
static int create_zlist_and_append (zlist_t **lp, void *item)
{
    if (!*lp && !(*lp = zlist_new()))
        return -1
    if (zlist_append (*lp, el) < 0)
        return -1;
    return 0;
}
```
Then just call
```c
if (create_zlist_and_append (&tmplist, el) < 0) {
    flux_log (h, LOG_ERR, "job-manager.disconnect: failed to remove event listener: out of memory");
    zlist_destroy (&tmplist);
    goto error;
}
```